website/docs/configuration/types.html.md

```diff
@@ -0,0 +1,187 @@
+grouped together as a single value. Structural types require a _schema_ as an
+argument, to specify which types are allowed for which elements.
+
+The two kinds of structural type in the Terraform language are:
+
+* `object(...)`: a collection of named attributes that each have their own type.
+
+    The schema for object types is `{<KEY> = <TYPE>, <KEY> = <TYPE>, ...}` — a
+    pair of curly braces containing a comma-separated series of `<KEY> = <TYPE>`
+    pairs. Values that match the object type must contain _exactly_ the
+    specified keys (no more and no fewer), and the value for each key must
+    match the specified type for that key.
+* `tuple(...)`: a sequence of elements identified by consecutive whole
+  numbers starting with zero, where each element has its own type.
+
+    The schema for tuple types is `[<TYPE>, <TYPE>, ...]` — a pair of square
+    brackets containing a comma-separated series of types. Values that match the
+    tuple type must have _exactly_ the same number of elements (no more and no
+    fewer), and the value in each position must match the specified type for
+    that position.
+
+For example: an object type of `object({name=string,age=number})` would match a
+value like the following:
+
+```hcl
+{
+  name = "John"
+  age  = 52
+}
+```
+
+And a tuple type of `tuple([string, number, bool])` would match a value like the
+following:
+
+```hcl
+["a", 15, true]
+```
+
+### Complex Type Literals
+
+The Terraform language has literal expressions for creating tuple and object
+values, which are described in [Expressions: Literal
+Expressions](./expressions.html#literal-expressions) as "list/tuple" literals
+and "map/object" literals, respectively.
+
+Terraform does _not_ provide any way to directly represent lists, maps, or sets.
+However, due to the automatic conversion of complex types (described below), the
+difference between similar complex types is almost never relevant to a normal
+user, and most of the Terraform documentation conflates lists with tuples and
+maps with objects. The distinctions are only useful when restricting input
+values for a module or resource.
+
+### Conversion of Complex Types
+
+Similar kinds of complex types (list/tuple/set and map/object) can usually be
+used interchangeably within the Terraform language, and most of Terraform's
+documentation glosses over the differences between the kinds of complex type.
+This is due to two conversion behaviors:
+
+* Whenever possible, Terraform converts values between similar kinds of complex
+  types if the provided value is not the exact type requested. "Similar kinds"
+  is defined as follows:
+    * Objects and maps are similar.
+    * Tuples and lists are similar.
+    * Sets are _almost_ similar to both tuples and lists. When a value is
+      converted to a set, duplicate values are discarded and the ordering of
+      elements is lost. When a set is converted to another type, no particular
+      ordering of elements is guaranteed.
```

Comment:
right?

Reply:
This is accurate.

We _do_ guarantee ordering of `set(string)` in particular to be lexicographical ordering by the string values, but perhaps that special case is better noted somewhere else. (Not sure where, though.)

Reply:
This is probably the place for it. 👍🏼